CanDIG/federation_service
# Add the ability to send requests with "safe" checking, exiting early if the service info cannot be quickly communicated with

# Jira ticket 

[DIG-1708](https://candig.atlassian.net/browse/DIG-1708)

# Description 

This PR introduces the ability to send requests with the "safe" header before a fanout call, which checks which servers are currently communicable. This was prompted by the BC server taking a long time to serve any request while the UHN server was experiencing DNS issues.

The idea here is to shortcut away requests (before the usual 60s timeout) if a server is unable to serve a service-info request. This involves a heartbeat script that is constantly checking the uptime of connected servers, and a "safe" header that can be used in all frontend requests. If a server is detected to be down by the heartbeat script, all safe requests will subsequently ignore that server.

# To test

1. Build with this branch 
2. Run integration tests.
3. Add candig-dev as a server to federation. 
4. candig-dev should show up in `/app/federation/live_servers.txt`, and requests to fanout 
5. Go offline, wait a minute, OR delete the entry in the container's `/app/federation/live_servers.txt`. 
6. If you waited, candig-dev should no longer show up in `/app/federation/live_servers.txt`
7. Send a request like the following:
```
curl -v candig.docker.internal:5080/federation/v1/fanout
  -H "Content-Type: application/json"
  -H "Authorization: Bearer $TOKEN"
  -d '{ "method": "GET", "path": "service-info", "service": "query", "payload": {}, "safe": "True" }'
```
You should get a response like the following:
```
[
  {
    "location": {
      "name": "Local",
      "province": "ON",
      "province-code": "ca-on"
    },
    "message": "handle_server_request failed on internal-1, federation = false: Safe check abort: internal-1 is assumed to be down",
    "service": "query",
    "status": 500
  }
]
```

[DIG-1708]: https://candig.atlassian.net/browse/DIG-1708?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
